### PR TITLE
feat: Measure Metric extend repository result to differentiate

### PR DESCRIPTION
**Pull request by MaronHatoum**
## Changes
added logs that show the total number of update PRs created by Renovate The total number of update PRs created by Renovate were merged. 
repo stats whether a repo is onboarded versus it has merged at least one Renovate PR.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/finalise/index.ts [hunk 1]
```diff
@@ -15,4 +17,17 @@ export async function finaliseRepo(
     `Action Required: Fix Renovate Configuration`
   );
   await clearRenovateRefs();
+  const prList = await platform.getPrList();
+  if (
+    prList?.some(
+      (pr) =>
+        pr.state === 'merged' &&
+        (pr.title !== 'Configure Renovate' ||
+          pr.title !== config.onboardingPrTitle)
```
Comment (rarkins) [inline lib/workers/repository/finalise/index.ts:26]: This will always return true if config.onbaordingPrTitle is set

File: lib/workers/repository/finalise/index.ts [hunk 2]
```diff
@@ -15,4 +17,17 @@ export async function finaliseRepo(
...
+    )
+  ) {
+    logger.info('Repo is activated');
```
Comment (rarkins) [inline lib/workers/repository/finalise/index.ts:29]: Debug

File: lib/workers/repository/result.ts [hunk 3]
```diff
@@ -49,23 +56,30 @@ export function processResult(
   ];
   const enabledStatuses = [CONFIG_SECRETS_EXPOSED, CONFIG_VALIDATION];
   let status: ProcessStatus;
-  let enabled: boolean | undefined;
+  let repoEnabled: boolean | undefined;
```
Comment (rarkins) [inline lib/workers/repository/result.ts:59]: Why rename?

File: lib/workers/repository/finalise/repository-statistics.spec.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  mockedFunction,
+} from '../../../../test/util';
+import { logger } from '../../../logger';
+import { getPrCache } from '../../../modules/platform/github/pr';
+import { runRenovateRepoStats } from './repository-statistics';
+
+jest.mock('../../../modules/platform/github/pr');
+jest.mock('../../../util/http/github');
+
+const prCache = loadJsonFixture('./pr-cache.json');
```
Comment (rarkins) [inline lib/workers/repository/finalise/repository-statistics.spec.ts:14]: Why is this needed? Can't you mock platform.getPrList() instead?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/result.ts (hunk 3 above)
Comment (MaronHatoum) [reply lib/workers/repository/result.ts:59]: I renamed this parameter to make it clearer

File: lib/workers/repository/finalise/index.ts (hunk 1 above)
Comment (MaronHatoum) [reply lib/workers/repository/finalise/index.ts:26]: config.onboardingPrTitle should not contain 'Configure Renovate'?
and if pr.title contains dependency update pr the condition will return true 
but if pr.title contains onboarding pr the condition will return false

Comment (rarkins) [reply lib/workers/repository/finalise/index.ts:26]: If onboardingPrTitle is "Foo" then this statement will always return true and therefore the onboarding PR will be counted

File: lib/workers/repository/result.ts (hunk 3 above)
Comment (rarkins) [reply lib/workers/repository/result.ts:59]: Try to avoid refactoring combined with feature changing, as it makes regression faults harder to detect and PRs longer to review. Please revert

File: lib/workers/repository/finalise/index.ts (hunk 1 above)
Comment (MaronHatoum) [reply lib/workers/repository/finalise/index.ts:26]: Do you mean that we should remove the second part of the condition?

File: lib/workers/repository/finalise/index.ts [hunk 5]
```diff
@@ -15,4 +17,14 @@ export async function finaliseRepo(
...
   );
   await clearRenovateRefs();
+  const prList = await platform.getPrList();
+  if (
+    prList?.some(
+      (pr) => pr.state === 'merged' && pr.title !== 'Configure Renovate'
+    )
+  ) {
+    logger.debug('Repo is activated');
+    config.repoIsActivated = true;
```
Comment (viceice) [inline lib/workers/repository/finalise/index.ts:27]: SHouldn't we check if the repo has a config file instead, so we can count manually onboarded repos?

File: lib/workers/repository/finalise/repository-statistics.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+function initStats(): PrStats {
+  return {
+    total: 0,
+    open: 0,
+    closed: 0,
+    merged: 0,
+  };
+}
+
+// eslint-disable-next-line require-await,@typescript-eslint/require-await
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.ts:22]: nope

File: lib/workers/repository/finalise/repository-statistics.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      case PrState.Merged:
+        prStats.merged += 1;
+        break;
+      case PrState.Closed:
+        prStats.closed += 1;
+        break;
+      case PrState.Open:
+        prStats.open += 1;
+        break;
+      // istanbul ignore next: exclude PrState.All & .NotOpen
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.ts:47]: nope, add a test instead

**Review by viceice — CHANGES_REQUESTED** [review]
please add some PR description about the intention

File: lib/workers/repository/finalise/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/workers/repository/finalise/index.ts]: No

Comment (MaronHatoum) [reply lib/workers/repository/finalise/index.ts]:
```
if (
    prList?.some(
      (pr) => pr.state === 'merged' && pr.title !== 'Configure Renovate'
    )
  ) 
```
this condition is not enough to filter onboarding PR?

File: lib/workers/repository/finalise/index.ts (hunk 5 above)
Comment (MaronHatoum) [reply lib/workers/repository/finalise/index.ts:27]: I think it is not enough to check config file only because we need to make sure that we have at least one merged PR by Renovate and this PR is not the onboarding one.

File: lib/workers/repository/result.ts [hunk 8]
```diff
@@ -67,5 +78,8 @@ export function processResult(
     logger.debug({ res }, 'Unknown res');
     status = 'unknown';
   }
-  return { res, status, enabled, onboarded };
+  logger.debug(
+    `Repository result: ${res} status: ${status} enabled: ${enabled} onboarded: ${onboarded}`
```
Comment (rarkins) [inline lib/workers/repository/result.ts:82]:
```suggestion
    `Repository result: ${res}, status: ${status}, enabled: ${enabled}, onboarded: ${onboarded}`
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/finalise/index.ts [hunk 9]
```diff
@@ -15,4 +17,14 @@ export async function finaliseRepo(
     `Action Required: Fix Renovate Configuration`
   );
   await clearRenovateRefs();
+  const prList = await platform.getPrList();
+  if (
+    prList?.some(
+      (pr) => pr.state === 'merged' && pr.title !== 'Configure Renovate'
```
[the remark this answers is not in the document]
Comment (MaronHatoum) [reply lib/workers/repository/finalise/index.ts]: I changed the condition to : 
```

  if (
    prList?.some(
      (pr) =>
        pr.state === 'merged' &&
        pr.title !== 'Configure Renovate' &&
        pr.title !== config.onboardingPrTitle
    )
  ) 
```
I don't know why it is not appear?

File: lib/workers/repository/finalise/__fixtures__/pr-list.json [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    "number": 1,
+    "sourceBranch": "renovate/configure",
+    "title": "Configure Renovate",
+    "state": "open",
+    "body": "Welcome to [Renovate](https://togithub.com/renovatebot/renovate)! This is an onboarding PR to help you understand and configure settings before regular Pull Requests begin.\n\n🚦 Renovate will begin keeping your dependencies up-to-date only once you merge or close this Pull Request.\n\n\n\n---\n### Detected Package Files\n\n * `Dockerfile` (dockerfile)\n\n### What to Expect\n\nWith your current configuration, Renovate will create 1 Pull Request:\n\n<details>\n<summary>Update dependency ubuntu to v22</summary>\n\n  - Schedule: [\"at any time\"]\n  - Branch name: `renovate/ubuntu-22.x`\n  - Merge into: `main`\n  - Upgrade ubuntu to `22.04`\n\n\n</details>\n\n\n---\n\n❓ Got questions? Check out Renovate's [Docs](https://docs.renovatebot.com/), particularly the Getting Started section.\nIf you need any further assistance then you can also [request help here](https://togithub.com/renovatebot/renovate/discussions).\n\n\n---\n\nThis PR has been generated by [Renovate Bot](https://togithub.com/renovatebot/renovate).\n",
```
Comment (viceice) [inline lib/workers/repository/finalise/__fixtures__/pr-list.json:8]: please inline this and remove all unused properties

File: lib/workers/repository/finalise/index.ts [hunk 11]
```diff
@@ -15,4 +17,17 @@ export async function finaliseRepo(
...
+  if (
+    prList?.some(
+      (pr) =>
+        pr.state === 'merged' &&
+        pr.title !== 'Configure Renovate' &&
+        pr.title !== config.onboardingPrTitle
+    )
+  ) {
+    logger.debug('Repo is activated');
+    config.repoIsActivated = true;
```
Comment (viceice) [inline lib/workers/repository/finalise/index.ts:30]: @rarkins I think this should use memcache (like the http stats) instead of changing our global config

https://github.com/renovatebot/renovate/blob/637585c5fd9f7d3239318ad9e9e3ec288ec3c6e1/lib/workers/repository/stats.ts#L7

File: lib/workers/repository/finalise/repository-statistics.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+function initStats(): PrStats {
+  return {
+    total: 0,
+    open: 0,
+    closed: 0,
+    merged: 0,
+  };
+}
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.ts:20]: this function doesn't make sense, so inline

File: lib/workers/repository/finalise/index.ts (hunk 11 above)
Comment (rarkins) [reply lib/workers/repository/finalise/index.ts:30]: At the moment it's inline with the existing config.repoIsOnboarded, so I'd prefer to leave it. The app reads the result/status of each run to determine if a repo is onboarded, and this is intended to let me see if the repo is truly active or they just onboarded and never merged a PR after that.

I agree we should look into a first class stats/metadata gathering concept, not necessarily in memCache but perhaps a module which is reset ever run

**Comment by MaronHatoum** [comment]
Hi @viceice 
can you see the last changes I pushed, please?
is this what you meant in inline?

File: lib/workers/repository/finalise/repository-statistics.spec.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+  mockedFunction,
+} from '../../../../test/util';
+import { logger } from '../../../logger';
+import { platform } from '../../../modules/platform';
+import { runRenovateRepoStats } from './repository-statistics';
+
+jest.mock('../../../modules/platform/github/pr');
+jest.mock('../../../util/http/github');
+
+const prList = loadJsonFixture('./pr-list.json');
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.spec.ts:14]:
```suggestion
const prList = Fixtures.get('./pr-list.json');
```

File: lib/workers/repository/finalise/repository-statistics.spec.ts [hunk 14]
```diff
@@ -0,0 +1,46 @@
+import {
```
Comment (viceice) [inline lib/workers/repository/finalise/repository-statistics.spec.ts:1]:
```suggestion
import { Fixtures } from '../../../../test/fixtures';
import {
```

File: lib/workers/repository/result.ts [hunk 15]
```diff
@@ -19,7 +20,13 @@ import {
 } from '../../constants/error-messages';
 import { logger } from '../../logger';
 
-type ProcessStatus = 'disabled' | 'enabled' | 'onboarding' | 'unknown';
+type ProcessStatus =
```
Comment (viceice) [inline lib/workers/repository/result.ts:23]:
```suggestion
export type ProcessStatus =
```

not your fault, but it's used in public interface, so should also be exported 😉

File: lib/workers/repository/finalise/repository-statistics.spec.ts (hunk 13 above)
Comment (viceice) [reply lib/workers/repository/finalise/repository-statistics.spec.ts]: 👀 Not yet Can you add a `getJson` function, which simply mirrors the old?

https://github.com/renovatebot/renovate/blob/5cf9123017dff463ed48bafcd7507c90065c850a/test/util.ts#L105-L111

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]
